pkg/messaging/kafka/publisher.go

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) Mainflux
+// SPDX-License-Identifier: Apache-2.0
+
+package kafka
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/mainflux/mainflux/pkg/messaging"
+	"github.com/segmentio/kafka-go"
+)
+
+var _ messaging.Publisher = (*publisher)(nil)
+
+type publisher struct {
+	url    string
+	mu     sync.Mutex
+	topics []string
+}
+
+// NewPublisher returns Kafka message Publisher.
+func NewPublisher(url string) (messaging.Publisher, error) {
+	ret := &publisher{
+		url: url,
+	}
+	return ret, nil
+
+}
+
+func (pub *publisher) Publish(topic string, msg messaging.Message) error {
+	if topic == "" {
+		return ErrEmptyTopic
+	}
+	data, err := proto.Marshal(&msg)
+	if err != nil {
+		return err
+	}
+	subject := fmt.Sprintf("%s.%s", chansPrefix, topic)
+	if msg.Subtopic != "" {
+		subject = fmt.Sprintf("%s.%s", subject, msg.Subtopic)
+	}
+
+	conn, err := kafka.Dial("tcp", pub.url)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+
+	topicConfigs := []kafka.TopicConfig{
+		{
+			Topic:             subject,
+			NumPartitions:     -1,
+			ReplicationFactor: -1,
+		},
+	}
+	if err := conn.CreateTopics(topicConfigs...); err != nil {
+		return err
+	}
+
+	writer := &kafka.Writer{
+		Addr:         kafka.TCP(pub.url),
+		Async:        true,
+		MaxAttempts:  100,
+		RequiredAcks: kafka.RequireAll,
+	}
+	defer writer.Close()
+
+	kafkaMsg := kafka.Message{
+		Value: data,
+		Topic: subject,
+	}
+	if err := writer.WriteMessages(context.Background(), kafkaMsg); err != nil {
+		return err
+	}
+	pub.mu.Lock()
+	defer pub.mu.Unlock()
+	if !pub.topicInTopics(subject) {
+		pub.topics = append(pub.topics, subject)
+	}
+	return nil
+}
+
+func (pub *publisher) Close() error {
```

Comment:
The publisher is closed only at the end of its lifetime - removing topics that way does not sound particularly useful since Publisher is meant to be long-lasting.

Reply:
What about cleaning up?

Reply:
I am not sure. Is there a way to configure Kafka to delete the topic if no client has pushed to it in a while (say 1d) or no client is subscribed? If we leave topics intact, how does that affect performance?

Reply:
From what I have reviewed I have not seen a way to automatically delete the topic after a specific period. That option is only done with message using [delete.retention.ms](https://kafka.apache.org/documentation/#topicconfigs_delete.retention.ms). Also, the number of topics isn't all that important in and of itself, but each Kafka topic is partitioned, and the total number of partitions may have an effect on performance [[1]](https://www.instaclustr.com/blog/the-power-of-kafka-partitions-how-to-get-the-most-out-of-your-kafka-cluster/).